clang/lib/CIR/CodeGen/CIRGenModule.cpp

```diff
@@ -678,7 +678,67 @@ mlir::cir::GlobalOp CIRGenModule::createGlobalOp(CIRGenModule &CGM,
 }
 
 void CIRGenModule::setCommonAttributes(GlobalDecl GD, mlir::Operation *GV) {
-  assert(!MissingFeatures::setCommonAttributes());
+  const Decl *D = GD.getDecl();
+  if (isa_and_nonnull<NamedDecl>(D))
+    setGVProperties(GV, dyn_cast<NamedDecl>(D));
+  else
+    assert(!MissingFeatures::setDefaultVisibility());
+
+  if (D && D->hasAttr<UsedAttr>())
+    assert(!MissingFeatures::addUsedOrCompilerUsedGlobal());
+
+  if (const auto *VD = dyn_cast_if_present<VarDecl>(D);
+      VD &&
+      ((codeGenOpts.KeepPersistentStorageVariables &&
+        (VD->getStorageDuration() == SD_Static ||
+         VD->getStorageDuration() == SD_Thread)) ||
+       (codeGenOpts.KeepStaticConsts && VD->getStorageDuration() == SD_Static &&
+        VD->getType().isConstQualified())))
+    assert(!MissingFeatures::addUsedOrCompilerUsedGlobal());
+}
+
+void CIRGenModule::setNonAliasAttributes(GlobalDecl GD, mlir::Operation *GO) {
+  const Decl *D = GD.getDecl();
+  setCommonAttributes(GD, GO);
+
+  if (D) {
+    auto GV = llvm::dyn_cast_or_null<mlir::cir::GlobalOp>(GO);
+    if (GV) {
+      if (D->hasAttr<RetainAttr>())
+        assert(!MissingFeatures::addUsedGlobal());
+      if (auto *SA = D->getAttr<PragmaClangBSSSectionAttr>())
+        assert(!MissingFeatures::addSectionAttributes());
+      if (auto *SA = D->getAttr<PragmaClangDataSectionAttr>())
+        assert(!MissingFeatures::addSectionAttributes());
+      if (auto *SA = D->getAttr<PragmaClangRodataSectionAttr>())
+        assert(!MissingFeatures::addSectionAttributes());
+      if (auto *SA = D->getAttr<PragmaClangRelroSectionAttr>())
+        assert(!MissingFeatures::addSectionAttributes());
+    }
+    auto F = llvm::dyn_cast_or_null<mlir::cir::FuncOp>(GO);
+    if (F) {
+      if (D->hasAttr<RetainAttr>())
+        assert(!MissingFeatures::addUsedGlobal());
+      if (auto *SA = D->getAttr<PragmaClangTextSectionAttr>())
+        if (!D->getAttr<SectionAttr>())
+          assert(!MissingFeatures::setSectionForFuncOp());
+
+      assert(!MissingFeatures::updateCPUAndFeaturesAttributes());
+    }
+
+    if (const auto *CSA = D->getAttr<CodeSegAttr>()) {
+      assert(!MissingFeatures::setSectionForFuncOp());
```

Comment:
@roro47 back to our discussion about sections, perhaps this could be related with the linker problem?